### PR TITLE
add `-Z dont-buffer-diagnostics`

### PR DESCRIPTION
**Pull request by pnkfelix**
Add `-Z dont-buffer-diagnostics`, a way to force NLL to immediately its diagnostics.

This is mainly intended for developers who want to see the error in its original context in the control flow. Two uses cases for that are:

  1. `-Z treat-err-as-bug` (which then allows extraction of a stack-trace to the origin of the error)

  2. RUST_LOG=... rustc, in which case it is often useful to see the logging statements that occurred immediately prior to the point where the diagnostic was signalled.

### CODE REVIEW

File: src/librustc/ty/context.rs [hunk 1]
```diff
@@ -1432,6 +1432,16 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         self.queries.on_disk_cache.serialize(self.global_tcx(), encoder)
     }
 
+    /// Either immediate signals the given diagnostic, or puts it on the
+    /// given `buffer` for later signalling in the future.
+    pub fn buffer(self, mut diag: DiagnosticBuilder, buffer: &mut Vec<Diagnostic>) {
+        if self.sess.opts.debugging_opts.dont_buffer_diagnostics {
```
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: I believe this should check `treat-err-as-bug` and abort. And/or you could probably make this:

```rust
if self.sess.opts.debugging_opts.treat_err_as_bug || self.sess.opts.debugging_opts.dont_buffer_diagnostics {
```

I've applied this patch myself numerous times.

Comment (nikomatsakis) [reply src/librustc/ty/context.rs]: Oh, wait, no I haven't. I think this function is in the wrong place. This code can live on `buffer` -- it has access to the [`Handler`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.Handler.html), which has the field `flags` (of type [`HandlerFlags`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.HandlerFlags.html)) that contain flags relevant to it. We can just add this flag in there.

File: src/librustc_mir/borrow_check/mutability_errors.rs [hunk 2]
```diff
@@ -457,7 +457,7 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
             }
         }
 
-        err.buffer(&mut self.errors_buffer);
+        self.tcx.buffer(err, &mut self.errors_buffer);
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mutability_errors.rs]: all these diffs aren't needed then

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc/ty/context.rs (hunk 1 above)
Comment (pnkfelix) [reply src/librustc/ty/context.rs]: Ok will do (probably on Monday)

**Comment by pnkfelix** [comment]
(this does not seem to yet work in the manner I intended.)

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**